src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp

```diff
@@ -166,12 +105,65 @@ void ParquetBlockOutputFormat::write(Chunk chunk, size_t row_group_size)
         file_writer = std::move(result.ValueOrDie());
     }
 
-    auto status = file_writer->WriteTable(*arrow_table, row_group_size);
+    // TODO: calculate row_group_size depending on a number of rows and table size
+    // allow slightly bigger than row_group_size to avoid a very small tail row group
+    auto status = file_writer->WriteTable(*arrow_table, std::max(format_settings.parquet.row_group_bytes, staging_bytes));
```

Comment:
Wait, doesn't WriteTable() take number of rows, not bytes?

you're right, I mixed them up.

Why max?

this is meant to replace https://github.com/ClickHouse/ClickHouse/blob/eb3b774ad048e3cb6db34511950c75df27679739/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp#L94


Would the TODO amount to just copy-pasting the num_row_groups and row_group_size lines from before the PR?

I'm not sure what does the TODO comment mean, I just kept it untouched

Reply:
> I'm not sure what does the TODO comment mean, I just kept it untouched

The TODO is added by this very PR, check [the commit](https://github.com/ClickHouse/ClickHouse/pull/50102/files) :) . It's probably an incorrectly resolved merge conflict: the TODO used to exist, then was removed in https://github.com/ClickHouse/ClickHouse/pull/49325 , then this PR is inadvertently bringing it back. You should probably un-remove the following code to replace the TODO:
```
    /// Increase row group size slightly (by < 2x) to avoid adding a small row groups for the
    /// remainder of the new chunk.
    /// E.g. suppose input chunks are 70K rows each, and max_rows = 1M. Then we'll have
    /// getNumRows() = 1.05M. We want to write all 1.05M as one row group instead of 1M and 0.05M.
    size_t num_row_groups = std::max(static_cast<UInt64>(1), accumulated_chunk.getNumRows() / target_rows);
    size_t row_group_size = (accumulated_chunk.getNumRows() - 1) / num_row_groups + 1; // round up
```

Reply:
the removed code is meant to "avoid adding a small row groups for the remainder of the new chunk.", IMHO this can be replaced by https://github.com/ClickHouse/ClickHouse/blob/565575cef9e5032d9424c8f649fa94a731bf8779/src/Processors/Formats/Impl/ParquetBlockOutputFormat.cpp#L118